lib/cn/kvset.h

```diff
@@ -83,6 +83,36 @@ enum {
     KVSET_MISS_KEY_TOO_LARGE = -2,
 };
 
+/**
+ * struct kvset_vgroup_map - vgroup map
+ *
+ * A vgroup map stores the number of vgroups and its corresponding index mappings
+ *
+ * The last vblock index from each vgroup is stored in vbidx_out.
+ *
+ * In the case of a split kvset where the kblocks are not rewritten, a source vblock
+ * index stored in its kblocks needs to be adjusted to obtain the correct output
+ * vblock index. This index adjust value is stored in vbidx_adj.
+ *
+ * vbidx_src is memory-resident and it exists purely for efficient vbidx conversion.
+ *
+ * The nvgroups, vbidx_out and vbidx_adj for each kvset are persisted in its hblock.
+ *
+ * Each kvset must contain a vgroup map. A vgroup map is established during all the
+ * different types of maintenance operations. However, queries and compaction
+ * operations consult a kvset's vgmap only if that kvset is a result of a split
+ * operation (flagged by setting a boolean in struct kvset).
+ *
+ * A vgroup map is also written for a kvset with zero vblocks with nvgroups as 0
+ * and w/o any vblock index mappings.
+ */
+struct kvset_vgroup_map {
+    uint32_t  nvgroups;    /* number of vgroups */
+    uint32_t *vbidx_out;   /* output index (indexes the vblock list in a kvset) */
+    uint32_t *vbidx_adj;   /* index adjust offset */
+    uint32_t *vbidx_src;   /* source index (vblock index recorded in the kblocks) */
+};
```

Comment:
Functions related to this struct are named `kvset_vgmap_alloc/free/etc`.  I'd be on board with changing struct name to `struct vgmap` and functions to `vgmap_alloc/free/etc`.

Reply:
I initially started in that direction but then thought it would make sense to have a _kvset__ prefix as it's a per kvset thing.. I'm up for removing the _kvset__ prefix from this struct and its related functions.